[#21] 개별 상품 정보 가져오기

ethan-sh-lee opened the pull request.

ethan-sh-lee committed:
기능 : 개별 상품 정보 가져오기

f-lab-dev commented on src/main/java/com/project/smartstore/service/ProductService.java:
```diff
@@ -22,4 +22,8 @@ public void addProductInStore(ProductDto productDto) {
     return productMapper.selectProductList(searchConditionDto);
   }
+
+  public ProductDto getProduct(int productId) {
```
이건 캐싱에 대해 고민해보면 좋을 것 같습니다.

1. 캐싱이란 무엇인지
2. Spring Cache에 대해 알아보시면 좋겠습니다~

ethan-sh-lee committed:
기타 : develop변경사항 merge

ethan-sh-lee committed:
기능 : EnableCaching 애노테이션과 cacheManager 설정 추가

ethan-sh-lee committed:
기능 : 개별 상품 가져오기에 캐싱 기능 추가

ethan-sh-lee committed:
기능 : getProduct sql문 수정

이전 작업에서 ProductDto에 tag와 registerDate 프로퍼티가 추가되었고 ProductMapper.xml에 getProduct sql문을 수정함

ethan-sh-lee committed:
기능 : 상품 수정하기 기능 추가

ethan-sh-lee committed:
기능 : 상품 수정하기에 캐시 수정 기능 추가

ethan-sh-lee committed:
기능 : 상품 삭제하기 기능 추가

ethan-sh-lee committed:
기능 : 상품 삭제하기 기능에 캐시 추가

상품 삭제하기 요청이 들어왔을 때 해당 상품에 대한 캐시가 있을경우 해당 캐시를 삭제하는 기능 추가

f-lab-dev commented on src/main/java/com/project/smartstore/config/RedisConfig.java:
```diff
@@ -54,4 +58,16 @@ public LettuceConnectionFactory redisConnectionFactory() {
+            RedisSerializationContext.SerializationPair.fromSerializer(new StringRedisSerializer()))
+        .serializeValuesWith(RedisSerializationContext.SerializationPair
+            .fromSerializer(new GenericJackson2JsonRedisSerializer()))
+        .entryTtl(Duration.ofMinutes(3L));
```
이렇게하면 모든 캐시 만료시간이 3초로 통일될 것 같은데 각각 다른 캐시 만료시간을 유연하게 지정할 수 있는 방법이 없을까요?

ethan-sh-lee replied:
cacheConfigurationMap에 캐시 이름과 원하는 캐시 설정을 따로 추가할 수 있도록 변경하였습니다.

ethan-sh-lee committed:
기능 : 레디스 설정 캐시별로 적용할 수 있도록 수정

기존에는 모든 캐시가 하나의 설정으로만 적용되도록하여 캐시별로 만료기간등을 동일하게 적용하여야했는데 캐시별 필요에 따라 설정을 변경할 수 있도록 수정하였습니다.
cacheConfigurationMap에 캐시 이름과 캐시 설정을 추가할 수 있다.

ethan-sh-lee committed:
기능 : 레디스 default 설정 누락된 것 추가

ethan-sh-lee committed:
스타일 : ProductMapper.xml 들여쓰기 수정

ethan-sh-lee committed:
Merge pull request #28 from f-lab-edu/feature/27

[#27] 상품 삭제하기 기능 추가

f-lab-dev commented on src/main/java/com/project/smartstore/config/RedisConfig.java:
```diff
@@ -60,14 +64,25 @@ public LettuceConnectionFactory redisConnectionFactory() {
-    RedisCacheConfiguration redisCacheConfiguration = RedisCacheConfiguration.defaultCacheConfig()
+  public CacheManager cacheManager(RedisConnectionFactory redisConnectionFactory) {
+    Map<String, RedisCacheConfiguration> cacheConfigurationMap = new HashMap<>();
+    cacheConfigurationMap.put("product",
```
`"product"` 이 값은 하드코딩하기보단 상수를 사용하는게 좋아보입니다~

ethan-sh-lee committed:
리팩토링 : 레디스 설정에 입력하는 하드코딩된 캐시네임을 enum 방식으로 변경

ethan-sh-lee committed:
리팩토링 : 캐시 어노테이션을 컨트롤러단에서 서비스단으로 이동

ethan-sh-lee committed:
Merge branch 'develop' of https://github.com/f-lab-edu/smart-store into feature/25

ethan-sh-lee committed:
Merge branch 'feature/25' of https://github.com/f-lab-edu/smart-store into feature/25

f-lab-dev commented on src/main/java/com/project/smartstore/constants/CacheNameConstants.java:
```diff
@@ -0,0 +1,7 @@
+
+public enum CacheNameConstants {
+
+  PRODUCT
```
enum과 상수중에 이거는 어떤걸 쓰는게 나을까요?

ethan-sh-lee replied:
음.. 저는 enum을 쓰는게 가독성과 타입 안정성에 도움이 된다고 생각하여 enum으로 진행하였습니다. 상수를 사용할 때 장점은 찾아보고 고민해봐도 잘모르겠습니다. 혹시 더 알아볼 키워드를 알려주실 수 있나요?

f-lab-dev replied:
이 경우는 문자열 상수가 필요한거로 타입이 명확하기 때문에 문자열 상수 형태로 두는게 어떨까요?

ethan-sh-lee committed:
Merge pull request #26 from f-lab-edu/feature/25

[#25] 상품 수정하기 기능 추가

ethan-sh-lee committed:
리팩토링 : 캐시 이름을 저장한 enum에서 상수 클래스로 변경

캐시 이름은 단순 문자열만을 필요로 하기 때문에 enum에서 상수를 저장한 class로 변경하고 캐시 애노테이션 또한 하드코딩에서 해당 상수 class를 이용하여 변경함
